lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp

```diff
@@ -635,12 +635,15 @@ class BroadcastOpConversionPattern
     mlir::Value input = srcOp.getOperand(0);
     mlir::Value result = srcOp.getResult();
 
-    if (srcOp->getUsers().empty()) {
-      return rewriter.notifyMatchFailure(
-          srcOp, "ttir.broadcast op should have at least one use.");
+    SmallVector<Operation *> srcOpUsers;
+    for (Operation *nextOp : srcOp->getUsers()) {
+      srcOpUsers.push_back(nextOp);
+    }
+    for (Operation *nextOp : srcOpUsers) {
+      rewriter.modifyOpInPlace(
+          srcOp, [&]() { nextOp->replaceUsesOfWith(result, input); });
```

Comment:
Does `rewriter.replaceAllOpUsesWith(nextOp, input);` not work? I prefer to use the more explicit methods from pattern rewriter because it'll a) guarantee do the right thing and b) opt into optimization cases where possible.

Reply:
No unfortunately it doesn't work, it keeps giving the following error: 

```
      * Fold {
      } -> FAILURE : unable to fold

      * Pattern : 'ttir.broadcast -> ()' {
      } -> FAILURE : pattern was already applied
    } -> FAILURE : no matched legalization pattern
    //===-------------------------------------------===//
  } -> FAILURE : failed to legalize operation updated in-place 'ttir.broadcast'
} -> FAILURE : pattern failed to match
} -> FAILURE : no matched legalization pattern
//===-------------------------------------------===//
../../test/ttmlir/Silicon/TTNN/simple_broadcast.mlir:19:11: error: failed to legalize operation 'ttir.broadcast'
```

However using `modifyInPlace` should be equivalent.

Reply:
Oh wait, can't this whole block

```c++
    SmallVector<Operation *> srcOpUsers;
    for (Operation *nextOp : srcOp->getUsers()) {
      srcOpUsers.push_back(nextOp);
    }
    for (Operation *nextOp : srcOpUsers) {
      rewriter.modifyOpInPlace(
          srcOp, [&]() { nextOp->replaceUsesOfWith(result, input); });
   }
```

Just be replaced with:
```c++
rewriter.replaceAllOpUsesWith(srcOp, input);
```

Reply:
It seems like the entire pattern matcher could read:
```c++
rewriter.replaceAllOpUsesWith(srcOp, srcOp.getOperand(0));
```

Reply:
@nsmithtt Yeah, but it just doesn't work for some weird reason when two broadcast ops are in the graph one after another, they end up not replaced in the graph and no error is thrown. Please read this for more info: https://github.com/tenstorrent/tt-mlir/pull/812#discussion_r1775628730

Reply:
One thought, maybe we can have a pass that detects chained broadcast ops and replace them with one broadcast op, where we calculate the target shape combining those original broadcast ops shapes.

Reply:
I read the following comment on a blog, I think the same applies to modification, that during rewrite, we can't modify operands of another op. That should explain this issue.

`"I think it should be mentioned in the (MLIR) doc that, during a rewrite, one can't call an erase on another op (an op other than the one matched) because such an op might still be part of the pattern matching worklist -- this is a common source of bugs. In contrast, the matched op has already been popped off the worklist and so it's safe to call erase on it."`

Reply:
IDK, we could consider having the CI docker set this flag